File: R/dnam_search_index.R
#!/usr/bin/env R

# Author: Sean Maden
#
# Functions to manage search index construction from DNAm array data, including
# dimension reduction with feature hashing and support for k nearest neighbors
# search lookup. The HNSW implementation in the hnswlib Python library is used
# for search index construction.
#

#' setup_sienv
#'
#' Set up a new virtual environment for search index construction using 
#' the basilisk package.
#' 
#' @param env.name Name of the new virtual environment (required, 
#' "dnam_si-hnswlib")
#' @param pkgv Vector of the dependencies and their versions for the new
#' virtual environment (required, format: "packagename==versionnum").
#' @returns New basilisk environment object.
#' @example
setup_sienv <- function(env.name = "dnam_si_hnswlib", 
                        pkgv = c("hnswlib==0.5.1", "pandas==1.2.2", 
                          "numpy==1.20.1", "mmh3==3.0.0", "h5py==3.2.1")){
  message("Defining the virtual env dependencies...")
  my_env <- basilisk::BasiliskEnvironment(envname = env.name, 
                                          pkgname = "recountmethylation", 
                                          packages = pkgv)
  message("Running virtual environment setup...")
  proc <- basilisk::basiliskStart(my_env) # define env process
  on.exit(basilisk::basiliskStop(proc)) # set exit process
  return(proc)
}

#' get_fh
#'
#' Get the hashed features for a data table. Uses reticulate package to
#' call the Python script to do feature hashing on a table of data. It is
#' assumed the input table has sample data in rows, with probe data in 
#' columns. The input data table should have row names but not column names.
#' 
#' @param csv_savepath Name/path of hashed features table to write (required,
#' string, writes new csv where rows = samples, cols = hashed features).
#' @param csv_openpath Name/path of table to hash (required, string, assumes 
#' a csv where rows = samples, cols = probes).
#' @param ndim Number of hashed features (integer, 1000).
#' @param lstart Line index to start on (0-based for Python, required, int, 0).
#' @returns Path to new hashed featuers table.
#' @examples
#' # get example bval csv
#' of_fpath <- system.file("extdata", "fhtest", 
#'          package = "recountmethylation")
#' of_fpath <- file.path(of_fpath, "tbval_test.csv")
#' # write new hashed features results
#' get_fh(csv_savepath = "bval_fn.csv", csv_openpath = of_fpath, ndim = 100)
#' @export
get_fh <- function(csv_savepath, csv_openpath, ndim = 1000, lstart = 1){
  message("Starting basilisk process...")
  proc <- recountmethylation:::setup_sienv()
  pyscript.path <- system.file("python", package = "recountmethylation")
  pyscript.path <- file.path(pyscript.path, "dnam_search_index.py")
  message("Doing feature hashing...")
  ncol <- length(unlist(strsplit(readLines(csv_openpath, 1), ",")))
  message("Target features will reduce dimensions from ", ncol, 
          " to ", ndim, "...")
  basilisk::basiliskRun(proc, function(pyscript.path, csv_savepath, 
                                       csv_openpath, ndim, lstart){
    message("Sourcing Python functions...")
    reticulate::source_python(pyscript.path)
    make_fhmatrix_autolabel(wf_name = csv_savepath, of_name = csv_openpath, 
                            ndim = as.integer(ndim), lstart = lstart)
  }, pyscript.path = pyscript.path, csv_savepath = csv_savepath, 
  csv_openpath = csv_openpath, ndim = ndim, lstart = lstart)
  if(file.exists(csv_savepath)){return(csv_savepath)} else{return(FALSE)}
  return(NULL)
}

#' make_si
#' 
#' Make search index from table of hashed features. Additional details about 
#' the hnswlib search index parameters (e..g `space_val`, `efc_val`, `m_val`, 
#' and `ef_val`) can be found in the Python package docstrings and ReadMe.
#' 
#' @param fh_csv_fpath Name/path of csv (e.g. a table of hashed features) 
#' containing data for the index (required, string, "bvaltest.csv", where 
#' rows = samples, cols = features).
#' @param si_fname Name of new search index file to save (required, string, 
#' "new_search_index.pickle")
#' @param si_dict_fname Name of new index dictionary, with string labels, to 
#' save (required, string, "new_index_dict.pickle").
#' @param threads Number of threads for processing new index (required, int, 4).
#' @param space_val Space value for new search index (required, valid string, 
#' l2').
#' @param efc_val EFC value for the index (required, int, 2000).
#' @param m_val M value for the index (required, int, 1000).
#' @param ef_val EF value for the index (required, int, 2000).
#' @returns Boolean, TRUE if new search index and dictionary created, FALSE if 
#' creating the new search index and dictionary files failed, otherwise NULL.
#' @example 
#' fh_csv_fpath <- system.file("extdata", "fhtest", 
#' package = "recountmethylation")
#' fh_csv_fpath <- file.path(fh_csv_fpath, "bval_fn.csv")
#' make_si(fh_csv_fpath)
#' @export
make_si <- function(fh_csv_fpath, si_fname = "new_search_index.pickle", 
                    si_dict_fname = "new_index_dict.pickle", threads = 4, 
                    space_val = 'l2', efc_val = 2000, m_val = 1000, ef_val = 2000){
  message("Starting basilisk process...")
  proc <- recountmethylation:::setup_sienv()
  pyscript.path <- system.file("python", package = "recountmethylation")
  pyscript.path <- file.path(pyscript.path, "dnam_search_index.py")
  basilisk::basiliskRun(proc, function(pyscript.path, fh_csv_fpath, si_fname, 
                                       si_dict_fname, threads, space_val, 
                                       efc_val, m_val, ef_val){
    message("Sourcing Python functions...")
    reticulate::source_python(pyscript.path)
    message("Making search index...")
    make_hnsw_si(fname = fh_csv_fpath, index_name = si_fname, 
            dindex_name = si_dict_fname, threads = as.integer(threads), 
            efc_val = as.integer(efc_val), m_val = as.integer(m_val), 
            ef_val = as.integer(ef_val))
  }, 
  pyscript.path = pyscript.path, fh_csv_fpath = fh_csv_fpath,
  si_fname = si_fname, si_dict_fname = si_dict_fname, threads = threads,
  space_val = space_val, efc_val = efc_val, m_val = m_val, ef_val = ef_val
  )
  if(file.exists(si_fname) & file.exists(si_dict_fname)){
    message("Made new search index '",si_fname,
            "' and search index dict '", si_dict_fname,"'")
    return(TRUE)
  } else{
    return(FALSE)
  }
  return(NULL)
}

#' query_si
#'
#' Query an HNSW search index. Does K Nearest Neighbors lookup on a previously 
#' saved search index object, returning the K nearest neighbors of the queried 
#' sample(s). The `query_si()` function returns verbose output, which can be 
#' silenced with suppressMessages()`.
#' 
#' @param sample_idv Vector of valid sample IDs, or GSM IDs, which are included
#' in the rownames of the hashed features table at fh_csv_fpath (requried, 
#' vector of char strings).
#' @param fh_csv_fpath Path to the hashed features table, which includes rownames
#' corresponding to sample ID strings in the sample_idv vector (required, char).
#' @param si_fname Base filename of the search index object, used to find the 
#' search index and index dict files, which are expected to be located at
#' si_fapth (required, char).
#' @param si_fpath Path to the directory containing the search index and index
#' dict files (required, char).
#' @param lkval Vector of K nearest neighbors to return per query (optional, 
#' int, c(1,2)).
#' @returns
#' @example
#' # file paths
#' # fh table
#' fh_csv_fname <- system.file("extdata", "fhtest", 
#' package = "recountmethylation")
#' fh_csv_fname <- file.path(fh_csv_fname, "bval_fh10.csv")
#' # si dict
#' index_dict_fname <- system.file("extdata", "sitest", 
#' package = "recountmethylation")
#' index_dict_fname <- file.path(index_dict_fname, "new_index_dict.pickle")
#' 
#' # set sample ids to query
#' sample_idv <- c("GSM1038308.1548799666.hlink.GSM1038308_5958154021_R01C01",
#'                 "GSM1038309.1548799666.hlink.GSM1038309_5958154021_R02C01")
#' # set a list of k nearest neighbors to query
#' lkval <- c(1,2,3)
#' 
#' # get query results as a data frame (with verbose results messaging)
#' dfk <- query_si(sample_idv = sample_idv, lkval = lkval, 
#'                 fh_csv_fname = "bval_fn.csv", 
#'                 index_dict_fname = "new_index_dict.pickle")
#' # returns:
#' # Starting basilisk process...
#' # Defining the virtual env dependencies...
#' # Running virtual environment setup...
#' # Sourcing Python functions...
#' # Querying the search index...
#' # Getting hashed features data for samples...
#' # Getting index data for sample: 
#' # GSM1038308.1548799666.hlink.GSM1038308_5958154021_R01C01'
#' # Getting index data for sample: 
#' # GSM1038309.1548799666.hlink.GSM1038309_5958154021_R02C01'
#' # Beginning queries of k neighbors from lk...
#' # ii =  0 , ki =  1
#' # Loading search index...
#' # Querying 2 elements in data with k = 1 nearest neighbors...
#' # Query completed, time: 0.0007359981536865234
#' # Applying labels to query results...
#' # Returning data (sample id, k index, and distance)...
#' # ii =  1 , ki =  2
#' # Loading search index...
#' # Querying 2 elements in data with k = 2 nearest neighbors...
#' # Query completed, time: 0.0006208419799804688
#' # Applying labels to query results...
#' # Returning data (sample id, k index, and distance)...
#' # ii =  2 , ki =  3
#' # Provided k '3' > n si samples, skipping...
#' # Returning query results...
#' @export
query_si <- function(sample_idv, fh_csv_fpath, 
                     si_fname = "new_search_index", 
                     si_fpath = ".", lkval = c(1,2)){
  message("Checking index and table locations...")
  if(!file.exists(fh_csv_fpath)){
    stop("Error: didn't find fh table at location:\n'",fh_csv_fpath,"'")}
  si_hnsw_fpath <- file.path(si_fpath, paste0(si_fname, ".pickle"))
  if(!file.exists(si_hnsw_fpath)){
    stop("Error: didn't find search index at location:\n'",v,"'")}
  si_dict_fpath <- file.path(si_fpath, paste0(si_fname, "_dict.pickle"))
  if(!file.exists(si_dict_fpath)){
    stop("Error: didn't find index dict at location:\n'",si_dict_fpath,"'")}
  message("Starting basilisk process...")
  proc <- recountmethylation:::setup_sienv()
  lkval <- lapply(lkval, as.integer) # format query k values list
  pyscript.path <- system.file("python", package = "recountmethylation")
  pyscript.path <- file.path(pyscript.path, "dnam_search_index.py")
  query_result <- basilisk::basiliskRun(proc, function(pyscript.path, 
    sample_idv, lkval, si_dict_fpath, fh_csv_fpath){
    message("Sourcing Python functions...")
    reticulate::source_python(pyscript.path)
    message("Querying the search index...")
    dfk <- make_dfk_sampleid(sample_idv = sample_idv, lk = lkval, 
      index_dict_fname = si_dict_fpath, fh_csv_fname = fh_csv_fpath)
    return(dfk)}, pyscript.path = pyscript.path, sample_idv = sample_idv, 
    lkval = lkval, si_dict_fpath = si_dict_fpath, fh_csv_fpath = fh_csv_fpath)
  return(query_result)
}

File: inst/python/dnam_search_index.py
#!/usr/bin/env python3

# Author: Sean Maden
#
# Manage DNAm array search indexes. Options to make an HNSW-based 
# search index from hashed features tables of DNAm array data.
# 

import mmh3
import numpy as np
import pandas as pd
import hnswlib, sys, os, re, pickle, random
from time import time
import faulthandler
faulthandler.enable()

def feature_hash(arr, target_dim=10000):
    """ Perform feature hashing on an array of data
    
    Perform feature hashing on the data in arr, into a vector of target_dim 
    total hashed features.

    Arguments:
        * arr: An array of values to be hashed.
        * target_dim: The target number of hashed values.
    Returns:
        * low_d_rep, or an array of hashed values of len == target_dim

    """ 
    low_d_rep = [0.0 for _ in range(target_dim)]
    for i, el in enumerate(arr):
        hashed = mmh3.hash(str(i))
        if hashed > 0.0:
            low_d_rep[hashed % target_dim] += arr[i]
        else:
            low_d_rep[hashed % target_dim] -= arr[i]
    return low_d_rep

def make_fhmatrix_autolabel(wf_name, of_name, lnotfloat = ['','NA','NaN'], 
    ndim = 10000, lstart = 0):
    """
    
    Get the hashed features table from an input data table. This function 
    automatically sets row labels as the first column of data in the input data
    table at `of_name.`

    Arguments:
        * wf_name: Name/path of hashed features table to write (required, 
            string, rows = samples, cols = hashed features).
        * of_name: Name/path of table to hash (required, tring, rows =  samples, 
            cols = probes).
        * lnotfloat: List of expected missing value symbols (required, ['','NA',
            'NaN']). These are replaced by the row-wise meidans of non-missing 
            values.
        * ndim: Number of hashed features (integer, 1000).
        * lstart: Line to start reading (required, int, 0).
    Returns:
        * None, saves new hashed features table to `wf_name`.

    """
    with open(of_name, "r") as fr:
        with open(wf_name, "w") as fw:
            for li, line in enumerate(fr):
                line_format = line.replace('\n', '').split(',')
                newrow = line_format[0]; lli = line_format[1::]
                if li >= lstart:
                    # replace NAs with median values
                    lli_median = np.median([float(ii) for ii in lli 
                        if not ii in lnotfloat])
                    lli_format = [float(ii) if not ii in lnotfloat
                                    else lli_median for ii in lli]
                    lli_fh = feature_hash(lli_format, target_dim = ndim)
                    newrow = newrow + ',' + ','.join([str(ii) for ii in lli_fh])
                    newrow = newrow + '\n'
                    print('Found new sample: '+newrow[0:100])
                    fw.write(newrow)    
                print("Finished with line number "+str(li))
    return None

def make_fhmatrix_specifylabels(labels_list, wf_name, of_name, 
    lnotfloat = ['','NA','NaN'], ndim = 10000):
    """ Make the hashed feaures matrix a Beta-values table

    Saves a .csv table of ndim total hashed features (columns) by elements, 
    where the number of elements is equal to the row count in the of_name table.
    Missing/NA values are replaced by the row median. New rows are processively
    written to the file wf_name.

    Arguments:
        * labels_list : Column names of the output table
        * wf_name : Name of the output table
        * of_name : Name of table to be hashed. Rows will be hashed. First line 
                    is colnames and thus skipped.
    Returns:
        * None, produces a new hashed features .csv table of dim nrow_of_name by 
        ndim+1 (columns), where first column has feature labels


    """
    with open(of_name, "r") as fr:
        with open(wf_name, "w") as fw:
            for li, line in enumerate(fr):
                if li > 0:
                    lli = line.replace('\n', '').split(',')[1::]
                    newrow = labels_list[li] # append label to new row
                    # replace NAs with median values
                    lli_median = np.median([float(ii) for ii in lli 
                        if not ii in lnotfloat])
                    lli_format = [float(ii) if not ii in lnotfloat
                                    else lli_median for ii in lli]
                    lli_fh = feature_hash(lli_format, target_dim = ndim)
                    newrow = newrow + ',' + ','.join([str(ii) for ii in lli_fh])
                    newrow = newrow + '\n'
                    print('Writing new row: '+newrow[0:100])
                    fw.write(newrow)
                    print("Finished with line number "+str(li))
    return None

def make_hnsw_si(fname, index_name, dindex_name, space_val = 'l2', threads = 10, 
    efc_val = 2000, m_val = 1000, ef_val = 2000):
    """ Makes a new search index from a csv table

    Arguments:
        * fname: Name of CSV file containing index data.
        * index_name: Name of new search index file to save.
        * dindex_name: Name of new index dict, with str labels, to save.
        * space_val: Space value for the index.
        * threads: Number of threads for processing the index.
        * efc_val: EFC value for the index.
        * m_val: M value for the index.
        * ef_val: EF value for the index.
    Returns
        * None, saves the new index and index dict

    """
    print("Loading the dataset...")
    df=pd.read_csv(fname, sep=',',header=None)
    df_str_labels = df.iloc[0::,0]
    df = df.iloc[0::,1::] # subset to exclude string labels
    num_elements = df.shape[0]
    dim = df.shape[1]
    print("Making the new index...")
    p_iter = hnswlib.Index(space = space_val, dim = dim)
    p_iter.set_num_threads(threads)
    df_labels = np.arange(num_elements)
    p_iter.init_index(max_elements = num_elements, 
        ef_construction = efc_val, M = m_val)
    p_iter.set_ef(ef_val)
    print("Adding new data to the search index...")
    p_iter.add_items(df, df_labels)
    print("Saving the search index...")
    p_iter.save_index(index_name)
    print("Saving the index dict with str labels, as pickle object...")
    dindex = {'index' : p_iter, 'strlabels' : df_str_labels}
    file_open = open(dindex_name, "wb")
    pickle.dump(obj = dindex, file = file_open)
    return None

def query_si(query_data, si_fname, si_labels = [], kval = 2):
    """ Perform a lookup on an hnswlib-saved search index

    Arguments:
        * query_data : Vector of feature hashed data, of dims R x C, where C is the 
                same as in siobject/search index object.
        * si_fname : Name/path to the queried search index.
        * si_labels : Sample ID labels corresponding to index positions in 
            search index at `si_fname`.
        * kval : K number of nearest neighbors to return from siobject lookup.

    Returns:
        * Vector of elements, where elements are indices if si_labels = False, 
                or else element labels if si_labels is provided.

    """
    time_start = time()
    print("Loading search index...")
    sidict = pickle.load(open(si_fname, "rb"))
    siobject = sidict["index"]
    print("Querying "+str(query_data.shape[0])+" elements in data with k = "+
        str(kval)+" nearest neighbors...")
    kval_labels, kval_distances = siobject.knn_query(query_data, k = kval)
    time_str = str(time()-time_start);print("Query completed, time: "+time_str)
    if len(si_labels) > 0:
        print("Applying labels to query results..."); kval_str_labels = []
        for ll in kval_labels:
            kval_str_labels.append([si_labels[ii] for ii in ll])
        print("Returning data (sample id, k index, and distance)...")
        return kval_str_labels, kval_labels, kval_distances
    else:
        print("Returning data (k index and distance)...")
        return kval_labels, kval_distances
    return NULL

def make_dfk_sampleid(sample_idv, lk = [1,2], fh_csv_fname = "bval_100_fh10.csv",
    index_dict_fname = "new_index_dict.pickle"):
    """ make_dfk_sampleid

    Get the sample labels for the k nearest neighbors on a series of queries.
    Use a vector gsmv to identify samples for the query, from metadata.
    
    Arguments:
        * sample_idv : Vector or list of sample ID strings, corresponding to 
            sample ID labels in the rownames/first column of the hashed features 
            table at `fh_csv_fname`, which can correspond to sample names in the 
            queried search index (requried, list/vector of strings).
        * lk : List of k nearest neighbors to query (required, list of int 
            values, 1000).
        * index_dict_fname: Name/path of the search index file (required, 
            string).
        * fh_csv_fname: Name/path of the hashed features csv to read (required, 
            string, )
    Returns:
        * dfk_final

    """
    print("Getting hashed features data for samples...")
    fhdict = {}
    with open(fh_csv_fname, "r") as of:
        for line in of:
            lline = line.split(",")
            sample_id = lline[0].replace('"', '').split(".")[0]
            if sample_id in sample_idv:
                print("Getting index data for sample: '{0}'".format(sample_id))
                fhdict[sample_id] = [float(fhi.replace("\n", "")) 
                    for fhi in lline[1::]
                ]
    dfi = pd.DataFrame.from_dict(fhdict).T
    si_dict = pickle.load(open(index_dict_fname, "rb"))
    dim_si_keylabv = len([ii for ii in si_dict['strlabels']])
    dfk = pd.DataFrame()
    dfk["sample_id"] = fhdict.keys()
    for ii, ki in enumerate(lk):
        print("Querying ",ki," neighbors from lk...")
        if ki <= dim_si_keylabv:
            kval_str_labels, kval_labels, kval_distances = query_si(
                query_data = dfi, si_fname = index_dict_fname, 
                si_labels = si_dict["strlabels"], kval = ki
            )
            kli_format = [";".join(ii) for ii in kval_str_labels]
            dfki = pd.DataFrame(kli_format)
            dfk['k=' + str(ki)] = [ii for ii in dfki[0]]
        else:
            print("Provided k '{0}' > n si samples, skipping...".format(ki))
    print("Returning query results...")
    return dfk

def main():
    """
    """